internal/dag/walk.go

```diff
@@ -0,0 +1,453 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package dag
```

Comment:
In my opinion we should not include these unused dag functions, just to minimize the amount of imported code.  I know we may use it in the future but its also likely we'll have to debug this existing functionality, and extra unused code makes that harder.

Reply:
As we discussed, I understand the rationale for that simplification, it does reduce the drag on maintenance, and if we don't know if/when we'll use those functions, we can get rid of them, and re-introduce them later.
That said, given the code was imported as-is (with very minimal differences) from TF, I would imagine it is relevant to keep it as similar as possible, if only so that if a bug is found in either or an improvement is added later, we can adopt it easily, or if there are changes that we want, it may be easier to reason with if there are no significant differences between the two packages.
It's more of a maybe/potential though, in the current state, I can understand why reducing the amount of code introduced would be desirable.
Any thoughts @nywilken @devashish-patel ?